presto-main/src/main/java/com/facebook/presto/server/SessionFactory.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.SessionPropertyManager;
+import com.facebook.presto.security.AccessControl;
+import com.facebook.presto.spi.QueryId;
+import com.facebook.presto.spi.security.Identity;
+import com.facebook.presto.transaction.TransactionManager;
+
+public interface SessionFactory
+{
+    Identity getIdentity();
+
+    Session createSession(
+            QueryId queryId,
```

Comment:
I also had confusion about this when reading the code. After discussing with @dain, the pattern is more clear: this is creating a closure over the request information, so that we can delay creation/validation and thus create a failed query rather than just returning an HTTP error.

The name `SessionFactory` seems bad, though I can't think of a better one.